_user/rotrectangle/main.c

```diff
@@ -220,24 +209,22 @@ int main(void)
 	int ret;
 	char *stack;
 
-	stack = (char *)malloc(1024 * sizeof(char));
-
 	graph_adapter_t adapter = GRAPH_ANY;
 	graph_mode_t mode = GRAPH_DEFMODE;
 	graph_freq_t freq = GRAPH_DEFFREQ;
 	graph_t graph;
 
-	rotrectangle_t rotrectangle;
+	volatile rotrectangle_t rotrectangle;
```

Comment:
Honestly, I don't see the point of passing this structure around. Wouldn't it be much simpler to declare it as a global variable?

Reply:
Alek, the idea of passing structure is better then define it globally because you can easily extend this program with threads rotating many rectangles on the same screen

Reply:
Ok, I was actually wondering if this is a motivation behind this. So only fix I see here is to make only some members of `rotrectangle_t` `volatile`, not whole `struct`.

Reply:
Oh, I thought we will go with lock-less programming, thus suggested `volatile`.

I'm not sure if our mutex'es are written correctly (if acquire/release operations include necessary memory barriers), but with them `volatile` should not be needed (see explanations in linux kernel docs: https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/tree/Documentation/process/volatile-considered-harmful.rst).

Please note that You missed `volatile` on `stop` which is shared between the threads.

Side note: I'm only reading the code, but I wonder if it's possible to slow down the render thread by writing may chars at once, maybe input handling thread should have lower priority?